pkg/iotool/timeoutwriter.go

```diff
@@ -0,0 +1,56 @@
+package iotool
+
+import (
+	"fmt"
+	"io"
+	"sync"
+	"time"
+)
+
+/*
+	TimeoutWriter accepts input until expiration, then returns errors.
+
+	An error will be written to the wrapped writer unless `Finished` is called
+	before the timeout.
+*/
+type TimeoutWriter struct {
+	w    io.Writer
+	mtx  sync.Mutex
+	done bool
+}
+
+var TimeoutErr = fmt.Errorf("writer closed due to timeout")
+
+func NewTimeoutWriter(w io.Writer, patience time.Duration) *TimeoutWriter {
+	return NewTimeoutWriterFromChan(w, time.After(patience))
+}
+
+func NewTimeoutWriterFromChan(w io.Writer, done <-chan time.Time) *TimeoutWriter {
+	t := &TimeoutWriter{w: w}
+	go func() {
+		<-done
+		t.mtx.Lock()
+		defer t.mtx.Unlock()
+		t.done = true
+		if w != nil {
+			fmt.Fprintf(w, "%s\n", TimeoutErr)
```

Comment:
Could just do `Fprintln`. Do we want some tag to identify the writer that timed out?